### PR TITLE
[17.01] History-local serial workflow scheduling.

### PR DESCRIPTION
**Pull request by dannon**
A less intrusive (and more easily removed, once we no longer need it) approach to resolving https://github.com/galaxyproject/galaxy/issues/3474.

@bgruening Want to test this with your use-case?

### CODE REVIEW

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by jgoecks** [comment]
Why make this an option? Seems this should be the one-and-only behavior.

**Comment by jmchilton** [comment]
@jgoecks This really breaks down for workflows that cannot be fully scheduled at one time - for instance workflows that use collection operations or have steps that produce dynamically discovered collections. Multiple sites have been using workflows of this style for at least a year and possibly longer and these things would not schedule properly with this in place.

**Comment by dannon** [comment]
@jgoecks I'd prefer it as the default configuration (so, enabled by default, for now), with an option to disable it, but I figured this way would have less pushback.

**Comment by dannon** [comment]
(looks like the failure here is just travis's osx box not being healthy)

**Comment by martenson** [comment]
@dannon correct, mac builds are hosed https://www.traviscistatus.com/incidents/k79mjcv403c4

**Comment by guerler** [comment]
This looks like a good fix to me. Thx @dannon.

**Comment by bgruening** [comment]
Works for me! Tested and I got the expected results! @guerler was this a +1?

**Comment by martenson** [comment]
I did merge this as the UX hurdle seems to be significant in some cases. However I would also like to immediately deprecate this config option in 17.01 release notes because this solution is far from ideal.

**Comment by dannon** [comment]
I think 'immediately deprecating' the option is a bit silly, and the case about this being the 'wrong' and 'broken' approach is pretty overstated.

There are improvements we can make down the road, but this makes things work like people expect in almost all cases.

**Comment by martenson** [comment]
Afaik with this option we are downgrading our own capabilities in addition to allowing use cases where invocations can go to limbo without any feedback to the user. I think that is a wrong approach to solving a UX hurdle. Given you switched the default to False I was thinking you saw it similarly.

**Comment by dannon** [comment]
No, I just figured it'd be easier to get it in and fix the problem for users having trouble with this, with less argument, if I swapped the default :)  I actually *would* like the default to be true, though, like I first wrote it.

I *really* do not see this as 'downgrading capabilities' for (to make up a statistic, which I think is reasonably accurate) 99.8% of people.  For the other .2%, people can execute workflows en-masse targetting different histories to execute in parallel if they would like.  A ton of workflows executing into the same history, given what we know about being able to use large histories, is a very bad idea right now anyway.

TLDR, the only people I think this will negatively affect 1) are very few in number, 2) have simple workarounds, 3) those workarounds (multiple histories) might be a better organizational solution anyway.

**Comment by martenson** [comment]
@dannon Gotcha, thanks for writing down your thinking.

**Comment by jmchilton** [comment]
>  I actually would like the default to be true, though, like I first wrote it.

I don't agree with the rest at all - but I truly appreciate the compromise on the default - thank you! It has been targeting the API since 16.10 and only @bgruening has complained. If more users find this problematic over time - I'll open the PR to switching the default myself.
